File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res: List[List[int]] = []
        nums.sort()
        for i in range(len(nums) - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            target = -nums[i]
            front = i + 1
            back = len(nums) - 1
            while front < back:
                summation = nums[front] + nums[back]
                if summation < target:
                    front += 1
                elif summation > target:
                    back -= 1
                else:
                    res.append([nums[i], nums[front], nums[back]])
                    while front < back and nums[front] == nums[front + 1]:
                        front += 1
                    while front < back and nums[back] == nums[back - 1]:
                        back -= 1
                    front += 1
                    back -= 1
        return res

File: LeetCode/0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(nums);
        for(int i=0;i<nums.length;i++)
        {
            if(i>0 && nums[i]==nums[i-1])
            {
                continue;
            }
            int target = -nums[i];
            int front = i+1;
            int back = nums.length - 1;
            while(front<back)
            {
                int sum = nums[front] + nums[back];
                if(sum<target)
                {
                    front++;
                }
                else if(sum>target)
                {
                    back--;
                }
                else
                {
                    res.add(Arrays.asList(nums[i],nums[front],nums[back]));
                    while(front<back && nums[front]==nums[front+1])
                    {
                        front++;
                    }
                    while(front<back && nums[back]==nums[back-1])
                    {
                        back--;
                    }
                    front++;back--;
                }
            }
        }
        return res;
    }
}